nep-12.mediawiki

```diff
@@ -0,0 +1,2098 @@
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request
+* Validate and format the inputs
+** Script hashes is valid
+** Format parameters
+*** If `byteArray`, convert to appropriate hexstring based on assumed format
+* Batch invoke inputs into script, and handle assets to each invoke as outputs
+* Set script transaction attribute 0x20 according to the following conditions
+** If triggerContractVerification is set to true, set 0x20 to scriptHash of the contract being invoked
+** If there is no fee, attachedAssets, or 'assetIntentOverrides', set 0x20 to the users address
+** If there are assetIntentOverrides but none of the inputs belong to the user address, set 0x20 to user address
+* If the "broadcastOverride" input argument is set to True, return the serialized signed transaction to the dApp
+* Else broadcast the RPC request
+* Return the transaction Id and the url of the rpc node which the tx was broadcast to
+
+Example
+<pre>
+{
+  txid: 'ed54fb38dff371be6e3f96e4880405758c07fe6dd1295eb136fe15f311e9ff77',
+  nodeUrl: 'http://seed7.ngd.network:10332',
+}:
+</pre>
+
+
+====signMessage====
+Signs a provided messaged with an account selected by user. A randomized salt prefix is added to the input string before it is signed, and it is encased in a non-executable transaction before signed. This ensures allow compatibility with Ledger devices.
+
+
+'''Method Interface'''
+
+<pre>
+function signMessage({message: string}): Promise<SignedMessage>
+</pre>
+
+
+'''Input arguments'''
+
+<pre>
+interface SignMessageArgs {
+  message: string; // Arbitrary message to sign
+}
+</pre>
+
+
+'''Success return value'''
+
+<pre>
+interface SignedMessage {
+  publicKey: string; // Public key of account that signed message
+  message: string; // Original message signed
+  salt: string; // Salt added to original message as prefix, before signing
+  data; string; // Signed message
+}
+</pre>
+
+
+'''Error return value'''
+
+<pre>
+interface Error {
+  type: string;
+  description: string;
+  data: string;
+}
+</pre>
+
+
+'''Example'''
+
+<pre>
+signMessage({
+  message: 'Hello World!',
+})
+.then((signedMessage: SignedMessage) => {
+  const {
+    publicKey,
+    message,
+    salt,
+    data,
+  } = signedMessage;
+
+  console.log('Public key used to sign:', publicKey);
+  console.log('Original message:', message);
+  console.log('Salt added to message:', salt);
+  console.log('Signed data:', data);
+})
+.catch(({type: string, description: string, data: any}) => {
+  switch(type) {
+    case UNKNOWN_ERROR:
+      console.log(description);
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request
+* Present the message to the user for signing
+* Create a randomized salt value to be prefixed to the message before signing
+* Create the hexstring to be signed. '010001f0' + {length of salt+message in hex} + salt + message + '0000'
```

Comment:
What is the salt?

Reply:
the salt is an arbitrary string added before signing so that the output signature will differ on every signature.